src/io/ozlibstream.cpp

```diff
@@ -91,16 +91,33 @@ int deflate_streambuf::sync()
 
 void ozlibstream::close()
 {
+    // Capture the exception mask so we can restore it if close() fails.
+    std::ios_base::iostate old_ex = exceptions();
+    bool close_failed = false;
     try
     {
         buf.close();
     }
     catch(...)
     {
-        setstate(badbit); //FIXME: This will throw the wrong type of exception
-                          //but there's no good way of setting the badbit
-                          //without causing an exception when exceptions is set
+        close_failed = true;
     }
+    if(!close_failed)
+        return;
+
+    // Setting the stream state while exceptions are enabled may cause
+    // `setstate` to throw an `ios_base::failure` which would replace
+    // the original exception. Temporarily disable exceptions on this
+    // stream, set the `badbit`, then restore the exception mask.
+    try {
+        exceptions(std::ios_base::goodbit);
```

Comment:
as, trial as asked before, please isolate try/catch only to lines that can throw. remove this line to above the block